aptos-move/block-executor/src/scheduler.rs

```diff
@@ -186,15 +186,21 @@ impl ValidationStatus {
     }
 }
 
-pub struct Scheduler {
-    /// Number of txns to execute, immutable.
-    num_txns: TxnIndex,
+impl Default for ValidationStatus {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct Scheduler<P: SchedulerProvider> {
+    provider: Arc<P>,
 
     /// An index i maps to indices of other transactions that depend on transaction i, i.e. they
     /// should be re-executed once transaction i's next incarnation finishes.
-    txn_dependency: Vec<CachePadded<Mutex<Vec<TxnIndex>>>>,
+    txn_dependency: P::TxnDependencyCollection,
+
     /// An index i maps to the most up-to-date status of transaction i.
-    txn_status: Vec<CachePadded<(RwLock<ExecutionStatus>, RwLock<ValidationStatus>)>>,
+    txn_status: P::TxnStatusCollection,
```

Comment:
@zekun000 i guess this is over-engineering to you?

Reply:
exactly, I don't think we need to template on the data structure but just uses vec. we could template on the index mapping if we want to keep exact the same current behavior.

we can define an index mapping trait for lookups
```
trait IndexLookup {
  fn lens() -> usize;
  fn positions(idx: TxnIndex) -> TxnIndex;
}
```

wdyt?

Reply:
lol, i think @sitalkedia was trying to pull me to the opposite way: to let the provider hold the `txn_status` collection.

Reply:
you're both too fancy

Reply:
@zjma - I think with @zekun000 's recent suggestion, we don't need any `trait` to start with, we can just make changes to default behavior. Hopefully that will not introduce any noticeable regression to default behavior.  If we see any significant regression, we can introduce this index mapping trait.